File: ui/src/services/engineStorageService.ts
import { constants } from '../constants';
import { BaseApi } from './api/baseApi';
import { type EngineImportData } from './engineDesignService';

export interface SavedEngineEntry {
  name: string;
  version: string;
  saved_at: string;
}

class EngineStorageService extends BaseApi {
  constructor() {
    super(constants.API_BASE_URL);
  }

  listEngines(): Promise<SavedEngineEntry[]> {
    return this.get<SavedEngineEntry[]>('/api/v1/engines');
  }

  saveEngine(name: string, version: string, data: EngineImportData): Promise<SavedEngineEntry> {
    return this.post<SavedEngineEntry>('/api/v1/engines', { ...data, name, version });
  }

  getEngine(name: string, version: string): Promise<EngineImportData> {
    return this.get<EngineImportData>(`/api/v1/engines/${encodeURIComponent(name)}/${encodeURIComponent(version)}`);
  }

  updateEngine(name: string, version: string, data: EngineImportData): Promise<SavedEngineEntry> {
    return this.put<SavedEngineEntry>(
      `/api/v1/engines/${encodeURIComponent(name)}/${encodeURIComponent(version)}`,
      { ...data, name, version },
    );
  }

  deleteEngine(name: string, version: string): Promise<{ message: string }> {
    return this.delete<{ message: string }>(
      `/api/v1/engines/${encodeURIComponent(name)}/${encodeURIComponent(version)}`,
    );
  }
}

export const engineStorageService = new EngineStorageService();

File: api/services/engine_storage_service.py
from services.storage.factory import get_storage_provider


def list_engines() -> list:
    return get_storage_provider().list_engines()


def save_engine(name: str, version: str, data: dict) -> dict:
    return get_storage_provider().save_engine(name, version, data)


def get_engine(name: str, version: str) -> dict:
    return get_storage_provider().get_engine(name, version)


def update_engine(name: str, version: str, data: dict) -> dict:
    return get_storage_provider().update_engine(name, version, data)


def delete_engine(name: str, version: str) -> None:
    get_storage_provider().delete_engine(name, version)
